src/librustc_codegen_llvm/lib.rs

```diff
@@ -308,7 +308,8 @@ impl CodegenBackend for LlvmCodegenBackend {
 
         sess.compile_status()?;
 
-        if !sess.opts.output_types.keys().any(|&i| i == OutputType::Exe ||
+        if sess.opts.debugging_opts.no_link ||
+           !sess.opts.output_types.keys().any(|&i| i == OutputType::Exe ||
                                                    i == OutputType::Metadata) {
             return Ok(());
```

Comment:
I noticed that `finalize_session_directory` is never called in this case, which means that incremental compilation won't work.

Reply:
Should it have been called for `Exe` and `Metadata` output types, too?

Reply:
The current condition says return if neither exe, nor metadata is emitted.

Reply:
I just realized that `codegen_backend.join_codegen_and_link` needs to be called to actually emit all object files. Maybe split rename it to `join_codegen` and return `CodegenResults` from it, leaving the linker part to `rustc_driver`?

Reply:
Good suggestion. In fact, I think it is a good refactoring even without this `-Z no-link` change.